src/Acuminator/Acuminator.Utilities/Roslyn/PXFieldAttributes/AcumaticaAttributesRelationsInfoProvider.cs

```diff
@@ -349,5 +352,61 @@ private static bool EqualsOrAggregatesAttributeDirectly(ITypeSymbol attribute, I
 		private static IEnumerable<ITypeSymbol> GetAllDeclaredAcumaticaAttributesOnClassHierarchy(ITypeSymbol type, PXContext pxContext) =>
 			type.GetAllAttributesDefinedOnThisAndBaseTypes()
 				.Where(attribute => attribute.IsDerivedFromPXEventSubscriberAttribute(pxContext));
+
+		public static CategorizedAttributes FilterTypeAttributes(this List<DacFieldAttributeInfo> attributesWithFieldTypeMetadata)
+		{
+			List<DacFieldAttributeInfo>? typeAttributesOnDacProperty = null;
+			List<DacFieldAttributeInfo>? typeAttributesWithDifferentDataTypesOnAggregator = null;
+			bool hasNonNullDataType = false;
+
+			foreach (var attribute in attributesWithFieldTypeMetadata)
+			{
+				var dataTypeAttributes = attribute.AggregatedAttributeMetadata
+												  .Where(atrMetadata => atrMetadata.IsFieldAttribute)
+												  .ToList();
+				if (dataTypeAttributes.Count == 0)
+					continue;
+
+				typeAttributesOnDacProperty ??= new List<DacFieldAttributeInfo>(capacity: 2);
+				typeAttributesOnDacProperty.Add(attribute);
+
+				if (dataTypeAttributes.Count == 1)
+				{
+					hasNonNullDataType = hasNonNullDataType || dataTypeAttributes[0].DataType != null;
+					continue;
+				}
+
+				int countOfDeclaredNonNullDataTypes = dataTypeAttributes.Where(atrMetadata => atrMetadata.DataType != null)
+																		.Select(atrMetadata => atrMetadata.DataType!)
+																		.Distinct<ITypeSymbol>(SymbolEqualityComparer.Default)
+																		.Count();
+				hasNonNullDataType = hasNonNullDataType || countOfDeclaredNonNullDataTypes > 0;
+
+				if (countOfDeclaredNonNullDataTypes > 1)
+				{
+					typeAttributesWithDifferentDataTypesOnAggregator ??= new List<DacFieldAttributeInfo>(capacity: 2);
+					typeAttributesWithDifferentDataTypesOnAggregator.Add(attribute);
+				}
+			}
+
+			return new(typeAttributesOnDacProperty, typeAttributesWithDifferentDataTypesOnAggregator, hasNonNullDataType);
+		}
+
+		public static TypesCompatibility CheckCompatibility(this DacPropertyInfo property, DacFieldAttributeInfo dataTypeAttribute)
+		{
+			var dataTypeFromAttribute = dataTypeAttribute.AggregatedAttributeMetadata
+														 .Where(atrMetadata => atrMetadata.IsFieldAttribute && atrMetadata.DataType != null)
+														 .Select(atrMetadata => atrMetadata.DataType!)
+														 .Distinct<ITypeSymbol>(SymbolEqualityComparer.Default)
+														 .FirstOrDefault();
+
```

Comment:
I don't like that this helper in general is not correct. This helper is public and it doesn't support out of the box `dataTypeAttribute` input that has incompatible data type attributes.

I understand that the passed `dataTypeAttribute` is already checked additionally in analyzers. This is a temporal coupling anti-pattern: https://blog.ploeh.dk/2011/05/24/DesignSmellTemporalCoupling/
 
I can think of three ways to fix this, the one you will choose is up to you:
- The simplest one is to make `CheckCompatibility` `internal` + write documentation for the method about additional expectations imposed on the `dataTypeAttribute` parameter

- Just copy paste this helper to both analyzers as a private implementation detail. I understand that it will violate DRY a bit, but IMO it's okay in this case. The most important thing is to use the same algorithm for the determination of compatibility of the property type with data type attributes. Originally, I didn't insist on making the analyzers call the same code, just to use the same algo. Perhaps, I didn't specify it clear enough, sorry for that. I do like the shared helper idea, but having a public helper with significant implicit constraints on the input outweighs the benefits for me.

- Add support for the case I described, so the helper will  work in all cases. I understand that it will be a little worse for performance, but it should be a non-existent difference. The code should be something like this:
```cs
public static TypesCompatibility CheckCompatibility(this DacPropertyInfo property, DacFieldAttributeInfo dataTypeAttribute)
{
	var distinctClrTypesFromDataTypeAttributes = 
		dataTypeAttribute.AggregatedAttributeMetadata
						 .Where(atrMetadata => atrMetadata.IsFieldAttribute && atrMetadata.DataType != null)
						 .Select(atrMetadata => atrMetadata.DataType!)
						 .ToHashSet<ITypeSymbol>(SymbolEqualityComparer.Default);

	switch (distinctClrTypesFromDataTypeAttributes.Count)
	{
		case 0:
			return TypesCompatibility.MissingTypeAnnotation;    //PXDBFieldAttribute and PXEntityAttribute without data type case
		case 1:
			var dataTypeFromAttributes = distinctClrTypesFromDataTypeAttributes.FirstOrDefault();

			if (dataTypeFromAttributes.Equals(property.PropertyTypeUnwrappedNullable, SymbolEqualityComparer.Default))
				return TypesCompatibility.CompatibleTypes;
			else
				return TypesCompatibility.IncompatibleTypes;

		default:
			return TypesCompatibility.IncompatibleTypes;	// Data type attributes correspond to more than one CLR type 
	}
}
```